src/Internal/RequestBuilder.php

```diff
@@ -0,0 +1,167 @@
+<?php
+declare(strict_types=1);
+
+namespace Retrofit\Internal;
+
+use GuzzleHttp\Psr7\MultipartStream;
+use GuzzleHttp\Psr7\Query;
+use GuzzleHttp\Psr7\Request;
+use GuzzleHttp\Psr7\Uri;
+use JetBrains\PhpStorm\ArrayShape;
+use Ouzo\Utilities\Arrays;
+use Ouzo\Utilities\Strings;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\StreamInterface;
+use Psr\Http\Message\UriInterface;
+use Retrofit\Attribute\HttpRequest;
+use Retrofit\Internal\Utils\Utils;
+use RuntimeException;
+
+class RequestBuilder
+{
+    private const PARAMETER_PLACEHOLDER = '{%s}';
+
+    private UriInterface $uri;
+    #[ArrayShape([0 => ['name' => 'string', 'value' => 'string']])]
+    private array $pathParameters = [];
+    private array $headers = [];
+    private array $fields = [];
+    private array $parts = [];
+    private ?StreamInterface $body = null;
+
+    /**
+     * @param array<string, string> $defaultHeaders
+     */
+    public function __construct(
+        UriInterface $baseUrl,
+        private readonly HttpRequest $httpRequest,
+        array $defaultHeaders = []
+    )
+    {
+        $this->uri = new Uri($baseUrl->__toString());
+        if (!is_null($this->httpRequest->path())) {
+            $this->uri = $this->uri->withPath($this->httpRequest->path());
+        }
+
+        foreach ($defaultHeaders as $name => $value) {
+            $this->addHeader($name, $value);
+        }
+    }
+
+    public function setBaseUrl(UriInterface|string $value): void
+    {
+        if (is_string($value)) {
+            $value = new Uri($value);
+        }
+        $this->uri = $value->withQuery($this->uri->getQuery());
+    }
+
+    public function addPathParam(string $name, string $value, bool $encoded): void
+    {
+        if (!$encoded) {
+            $value = rawurlencode($value);
+        }
+        $this->pathParameters[] = ['name' => $name, 'value' => $value];
+    }
+
+    private array $queries = [];
+
+    public function addQueryParam(string|array $name, string|array|null $value, bool $encoded): void
+    {
+        if (is_null($value)) {
+            $name = Arrays::toArray($name);
+            foreach ($name as $item) {
+                if (!$encoded) {
+                    $item = rawurlencode($item);
+                }
+                $this->queries[] = $item;
+            }
+        } else {
+            $value = Arrays::toArray($value);
+            foreach ($value as $item) {
```

Comment:
`Arrays::map`?

Reply:
In the previous case we can easily use `Arrays::map`, here it's not looks good to me.